client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java

```diff
@@ -93,17 +93,39 @@ public synchronized int register(String subject, Schema schema)
   }
 
   @Override
-  public synchronized Schema getByID(int id) throws IOException, RestClientException {
-    if (idCache.containsKey(id)) {
-      return idCache.get(id);
+  public synchronized Schema getBySubjectAndID(String subject, int id)
+      throws IOException, RestClientException {
+    Map<Integer, Schema> idSchemaMap;
+    if (idCache.containsKey(subject)) {
+      idSchemaMap = idCache.get(subject);
     } else {
+      idSchemaMap = new IdentityHashMap<Integer, Schema>();
+      idCache.put(subject, idSchemaMap);
+    }
+
+    if (idSchemaMap.containsKey(id)) {
+      return idSchemaMap.get(id);
+    } else {
+      if (idSchemaMap.size() >= identityMapCapacity) {
+        throw new IllegalStateException("Too many schema objects created for " + subject + "!");
+      }
       Schema schema = getSchemaByIdFromRegistry(id);
-      idCache.put(id, schema);
+      idSchemaMap.put(id, schema);
       return schema;
     }
   }
 
   @Override
+  public synchronized Schema getByID(int id) throws IOException, RestClientException {
+    for (Map<Integer, Schema> idSchemaMap : idCache.values()) {
```

Comment:
Could we simplify this by treating `getByID` as "subject-less", i.e. a `null` or `""` value for subject? I think this would make all the expected semantics work out:

1. When you're using a "normal" deserializer, you load it here and use an empty subject. This effectively puts it in its own caching namespace since it has its own entry in the map.
2. When you're using a Converter which will annotate the schema with the version, you'll be including the subject and so each subject gets its own caching namespace.

It would also a) avoid looping over multiple entries and b) I think the current version won't properly cache the result since `getSchemaByIdFromRegistry` so deserializing for a given ID repeatedly using a deserializer rather than converter will have to look up the schema in the registry repeatedly.